File: sdk/go/event_mesh.go
package alpgo

import (
	"fmt"
	"time"
)

type MeshEvent struct {
	ID          string
	Topic       string
	SenderAgent string
	Payload     string
	EventType   string
	Timestamp   string
}

func NewMeshEvent(id, topic, senderAgent, payload string) *MeshEvent {
	return &MeshEvent{
		ID:          id,
		Topic:       topic,
		SenderAgent: senderAgent,
		Payload:     payload,
		EventType:   "state_change",
		Timestamp:   time.Now().Format(time.RFC3339),
	}
}

type EventHandler func(event *MeshEvent)

type EventMeshEngine struct {
	subscriptions map[string][]EventHandler
	eventBuffer   []*MeshEvent
}

func NewEventMeshEngine() *EventMeshEngine {
	return &EventMeshEngine{
		subscriptions: make(map[string][]EventHandler),
		eventBuffer:   []*MeshEvent{},
	}
}

func (e *EventMeshEngine) Subscribe(topic string, handler EventHandler) {
	e.subscriptions[topic] = append(e.subscriptions[topic], handler)
}

func (e *EventMeshEngine) Publish(eventID, topic, senderAgent, payload string) *MeshEvent {
	event := NewMeshEvent(eventID, topic, senderAgent, payload)
	e.eventBuffer = append(e.eventBuffer, event)
	if handlers, ok := e.subscriptions[topic]; ok {
		for _, handler := range handlers {
			handler(event)
		}
	}
	return event
}

func (e *EventMeshEngine) EventBuffer() []*MeshEvent {
	return e.eventBuffer
}

func (e *EventMeshEngine) BufferLen() int {
	return len(e.eventBuffer)
}

func (e *EventMeshEngine) SubscriptionCount(topic string) int {
	return len(e.subscriptions[topic])
}

func init() {
	fmt.Printf("[alpgo] event_mesh initialized\n")
}

File: sdk/rust/src/event_mesh.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct MeshEvent {
    pub id: String,
    pub topic: String,
    pub sender_agent: String,
    pub payload: String,
    pub event_type: String,
    pub timestamp: String,
}

impl MeshEvent {
    pub fn new(
        id: impl Into<String>,
        topic: impl Into<String>,
        sender_agent: impl Into<String>,
        payload: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            topic: topic.into(),
            sender_agent: sender_agent.into(),
            payload: payload.into(),
            event_type: "state_change".into(),
            timestamp: chrono::Utc::now().to_rfc3339(),
        }
    }
}

pub type EventHandler = Box<dyn Fn(&MeshEvent) + Send + Sync>;

pub struct EventMeshEngine {
    subscriptions: HashMap<String, Vec<EventHandler>>,
    event_buffer: Vec<MeshEvent>,
}

impl EventMeshEngine {
    pub fn new() -> Self {
        Self {
            subscriptions: HashMap::new(),
            event_buffer: Vec::new(),
        }
    }

    pub fn subscribe<F>(&mut self, topic: impl Into<String>, handler: F)
    where
        F: Fn(&MeshEvent) + Send + Sync + 'static,
    {
        self.subscriptions
            .entry(topic.into())
            .or_default()
            .push(Box::new(handler));
    }

    pub fn publish(
        &mut self,
        event_id: impl Into<String>,
        topic: impl Into<String>,
        sender_agent: impl Into<String>,
        payload: impl Into<String>,
    ) -> MeshEvent {
        let event = MeshEvent::new(event_id, topic, sender_agent, payload);
        self.event_buffer.push(event.clone());
        if let Some(handlers) = self.subscriptions.get(&event.topic) {
            for handler in handlers {
                handler(&event);
            }
        }
        event
    }

    pub fn event_buffer(&self) -> &[MeshEvent] {
        &self.event_buffer
    }
}
